frame/transaction-payment/rpc/src/lib.rs

```diff
@@ -107,14 +109,35 @@ where
 				Some(format!("{:?}", e)),
 			))
 		})?;
-		api.query_info(&at, uxt, encoded_len).map_err(|e| {
+
+		fn map_err(error: impl ToString, desc: &'static str) -> CallError {
 			CallError::Custom(ErrorObject::owned(
 				Error::RuntimeError.into(),
-				"Unable to query dispatch info.",
-				Some(e.to_string()),
+				desc,
+				Some(error.to_string()),
 			))
-			.into()
-		})
+		}
+
+		if api
+			.api_version::<dyn TransactionPaymentRuntimeApi<Block, Balance>>(&at)
+			.map_err(|e| map_err(e, "Transaction payment runtime api not present."))?
+			.unwrap_or(0) <
+			2
+		{
+			#[allow(deprecated)]
+			api.query_info_before_version_2(&at, uxt, encoded_len)
+				.map_err(|e| map_err(e, "Unable to query dispatch info.").into())
+		} else {
+			let res = api
+				.query_info(&at, uxt, encoded_len)
+				.map_err(|e| map_err(e, "Unable to query dispatch info."))?;
+
+			Ok(RuntimeDispatchInfo {
+				weight: sp_weights::OldWeight(res.weight.ref_time()),
```

Comment:
I think this really should be using the new weight and not the old one, as it's already in V2.

Reply:
No. The RPC was always returning the old weight and we should stick to this. Otherwise this pr is useless, as downstream would still fail.

Reply:
I guess since RPCs don't have versioning like `state_call`, it makes sense to never update them.